### ISSUE TITLE
[Q]: Splitting multiclass masks

### ISSUE THREAD

**Issue by meption** [qubvel/segmentation_models#123]
I have 20 images where each image contains multiple cars/buses/trucks etc.
I also have 20 mask images where each mask contains one or more classes (0-5) (where 0 is background)

The mask labels require a `[N, H, W, C]` structure but my question is can I input 1 full mask image
into each array slot which would be then `y_train = [20, 224, 244, 6]` or do I need to split each class in each mask?

So if `mask1.png` would have 3 classes `[0, 2, 3]` I would have to do
```
y_train[0, :, :, 0] = bg_pixels
y_train[1, :, :, 2] = bus_pixels #from mask[mask == 2]
y_train[2, :, :, 3] = truck_pixels
```

Eventually resulting in `y_train = [84, 224, 244, 6]`?

Thanks in advance

**Comment by qubvel** [comment]
Hi @meption 
Yes, you have to prepare your masks to be onehot arrays. 
```python
y_train = np.zeros((n, h, w, 6), dtype='uint8')
for i, mask in enumerate(masks):
  for j, class_value in enumerate(class_values):
     y_train[i, ..., j] = (mask == class_value).astype('uint8')
```
Something like that.

**Comment by meption** [comment]
Sorry for opening but I'm not quite sure I follow in assignment of `y_train[i, ..., j]`.

Perhaps I've misunderstood something but `categorical` turns into binary vectors of 0's and1's, so `mask == class_value` would never match properly and `categorical` output is `(224, 224, 6)` to fill the `(H, W, C)` which results in a shape error, I presume `y_train[0, :, :, :]` is wrong as well

```
n_classes = 6
n_masks = 20
masks = []
for p in glob("data/train/masks/*.png"):
    image = imread(p, as_gray = 1)
    image = img_as_ubyte(resize(image, (h, w)))
    categorical = to_categorical(image, num_classes=n_classes, dtype="uint8")
    masks.append(cate)
    
y_train = np.zeros((n_masks, h, w, n_classes), dtype='uint8')
for i, mask in enumerate(masks):
    for j, class_value in enumerate(class_values):
        y_train[i, ..., j] = (mask == class_value).astype('uint8')
```

**Comment by qubvel** [comment]
1. For each image (H, W, C) should be only one mask with shape (H, W, N_CLASSES) and it should be categorical/onehot (each class contain only {0, 1} values)
2. Suppose you have .png masks where classes are encoded as unique values (e.g. 0, 1, 2, etc..) just read them and convert co categorical (you can use keras.utils.to_categorical OR script provided by me)
